// gitbook-styleguide/lib/Badge.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');

var STYLES = require('./STYLES');

var Badge = React.createClass({
    displayName: 'Badge',
    getDefaultProps: function getDefaultProps() {
        return {
            style: STYLES[0]
        };
    },
    render: function render() {
        var style = this.props.style;
        var className = classNames('badge', 'badge-' + style, this.props.className);

        return React.createElement(
            'span',
            { className: className },
            this.props.children
        );
    }
});

/**
 * Create a style for badges
 * @param {String} style
 * @return {React.Component}
 */
function createBadgeStyle(style) {
    return React.createClass({
        displayName: Badge.displayName + style,
        render: function render() {
            return React.createElement(Badge, _extends({}, this.props, { style: style.toLowerCase() }));
        }
    });
}

module.exports = Badge;
module.exports.Info = createBadgeStyle('Info');
module.exports.Danger = createBadgeStyle('Danger');
module.exports.Success = createBadgeStyle('Success');
module.exports.Warning = createBadgeStyle('Warning');
