// gitbook-styleguide/lib/Description.js
'use strict';

var React = require('react');

var Container = require('./Container');

/**
 * Component to create intros in PageHead
 */
var HeadIntro = React.createClass({
    displayName: 'HeadIntro',


    propTypes: {
        title: React.PropTypes.string,
        description: React.PropTypes.element,
        avatarUrl: React.PropTypes.string,
        extra: React.PropTypes.element // Extra info on the side
    },

    render: function render() {
        return React.createElement(
            'div',
            { className: 'head-intro' },
            React.createElement(
                Container,
                null,
                React.createElement(
                    'h1',
                    null,
                    this.props.title
                ),
                React.createElement(
                    'p',
                    { className: 'lead' },
                    this.props.description
                )
            )
        );
    }
});

module.exports = HeadIntro;
