// gitbook-styleguide/lib/FilterList.js
'use strict';

var React = require('react');
var classNames = require('classnames');

var FilterItem = React.createClass({
    displayName: 'FilterItem',
    getDefaultProps: function getDefaultProps() {
        return {
            selected: false,
            href: '#'
        };
    },
    onClick: function onClick(e) {
        if (!this.props.onClick) {
            return;
        }

        e.preventDefault();
        this.props.onClick();
    },
    render: function render() {
        var className = classNames('filter-item', this.props.className, {
            selected: this.props.selected
        });
        var count = this.props.count;
        var href = this.props.href;
        var inner = '';

        if (typeof count !== 'undefined') {
            inner = React.createElement(
                'span',
                { className: 'count' },
                count
            );
        }

        return React.createElement(
            'li',
            null,
            React.createElement(
                'a',
                { className: className, href: href, onClick: this.onClick },
                inner,
                this.props.children
            )
        );
    }
});

var FilterList = React.createClass({
    displayName: 'FilterList',
    render: function render() {
        return React.createElement(
            'ul',
            { className: 'filter-list' },
            this.props.children
        );
    }
});

module.exports = FilterList;
module.exports.Item = FilterItem;
