// gitbook-styleguide/lib/Form.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');

var Form = React.createClass({
    displayName: 'Form',
    submit: function submit() {
        this.refs.form.submit();
    },
    render: function render() {
        return React.createElement(
            'form',
            _extends({ ref: 'form' }, this.props),
            this.props.children
        );
    }
});

var FormGroup = React.createClass({
    displayName: 'FormGroup',
    getDefaulProps: function getDefaulProps() {
        return {
            error: false
        };
    },
    render: function render() {
        var className = classNames('form-group', {
            'has-error': this.props.error
        }, this.props.className);

        return React.createElement(
            'div',
            { className: className },
            this.props.children
        );
    }
});

var FormHelpBlock = React.createClass({
    displayName: 'FormHelpBlock',
    render: function render() {
        return React.createElement(
            'p',
            { className: 'help-block' },
            this.props.children
        );
    }
});

module.exports = Form;
module.exports.Group = FormGroup;
module.exports.HelpBlock = FormHelpBlock;
