// gitbook-styleguide/lib/Input.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');

var SIZES = require('./SIZES');

var Input = React.createClass({
    displayName: 'Input',


    /**
     * Focus this input
     */
    focus: function focus() {
        this.refs.input.focus();
    },
    getDefaultProps: function getDefaultProps() {
        return {
            type: 'text',
            size: SIZES[0]
        };
    },
    onChange: function onChange(e) {
        if (this.props.onChange) {
            this.props.onChange(e);
        }
    },
    render: function render() {
        var _props = this.props;
        var onBlur = _props.onBlur;
        var onFocus = _props.onFocus;
        var onKeyDown = _props.onKeyDown;
        var onClick = _props.onClick;
        var value = _props.value;
        var readOnly = _props.readOnly;
        var defaultValue = _props.defaultValue;
        var size = _props.size;
        var autoFocus = _props.autoFocus;
        var placeholder = _props.placeholder;
        var type = _props.type;
        var name = _props.name;
        var disabled = _props.disabled;


        var className = classNames('form-control', 'input-' + size, this.props.className);

        return React.createElement('input', { ref: 'input',
            type: type, autoFocus: autoFocus,
            className: className, disabled: disabled, readOnly: readOnly,
            name: name, value: value, defaultValue: defaultValue,
            placeholder: placeholder,
            onFocus: onFocus,
            onBlur: onBlur,
            onKeyDown: onKeyDown,
            onClick: onClick,
            onChange: this.onChange
        });
    }
});

/**
 * Create a style of alert
 * @param {String} style
 * @return {React.Component}
 */
function createInputType(type) {
    return React.createClass({
        displayName: Input.displayName + type,
        render: function render() {
            return React.createElement(Input, _extends({}, this.props, { type: type.toLowerCase() }));
        }
    });
}

var InputGroup = React.createClass({
    displayName: 'InputGroup',
    render: function render() {
        var className = classNames('input-group', this.props.className || []);
        return React.createElement(
            'div',
            { className: className },
            this.props.children
        );
    }
});

var InputGroupAddon = React.createClass({
    displayName: 'InputGroupAddon',
    render: function render() {
        return React.createElement(
            'span',
            _extends({}, this.props, { className: 'input-group-addon' }),
            this.props.children
        );
    }
});

module.exports = Input;
module.exports.Email = createInputType('Email');
module.exports.Password = createInputType('Password');
module.exports.Group = InputGroup;
module.exports.GroupAddon = InputGroupAddon;
