// gitbook-styleguide/lib/Menu.js
'use strict';

var React = require('react');
var classNames = require('classnames');

var MenuItem = React.createClass({
    displayName: 'MenuItem',
    getDefaultProps: function getDefaultProps() {
        return {
            active: false,
            href: null
        };
    },
    render: function render() {
        var inner = void 0;
        var className = classNames(this.props.className || '', {
            'active': this.props.active
        });

        if (this.props.href) {
            inner = React.createElement(
                'a',
                { href: this.props.href },
                this.props.children
            );
        } else {
            inner = React.createElement(
                'span',
                null,
                this.props.children
            );
        }

        return React.createElement(
            'li',
            { className: className },
            inner
        );
    }
});

var Menu = React.createClass({
    displayName: 'Menu',
    getDefaultProps: function getDefaultProps() {
        return {
            active: false,
            href: null
        };
    },
    render: function render() {
        var className = classNames('menu', {
            'pull-right': this.props.right,
            'pull-left': this.props.left
        });

        return React.createElement(
            'ul',
            { className: className },
            this.props.children
        );
    }
});

module.exports = Menu;
module.exports.Item = MenuItem;
