// gitbook-styleguide/lib/Textarea.js
'use strict';

var React = require('react');
var classNames = require('classnames');

var SIZES = require('./SIZES');

var Textarea = React.createClass({
    displayName: 'Textarea',


    /**
     * Focus this input
     */
    focus: function focus() {
        this.refs.textarea.focus();
    },
    getDefaultProps: function getDefaultProps() {
        return {
            size: SIZES[0],
            rows: 3
        };
    },
    onChange: function onChange(e) {
        if (this.props.onChange) {
            this.props.onChange(e);
        }
    },
    render: function render() {
        var _props = this.props;
        var disabled = _props.disabled;
        var name = _props.name;
        var placeholder = _props.placeholder;
        var size = _props.size;
        var rows = _props.rows;
        var autoFocus = _props.autoFocus;
        var value = _props.value;
        var defaultValue = _props.defaultValue;
        var maxLength = _props.maxLength;


        var className = classNames('form-control', 'input-' + size, this.props.className);

        return React.createElement('textarea', { ref: 'textarea',
            className: className, rows: rows, maxLength: maxLength,
            autoFocus: autoFocus, disabled: disabled,
            name: name, value: value, defaultValue: defaultValue,
            placeholder: placeholder, onChange: this.onChange });
    }
});

module.exports = Textarea;
