// gitbook-plugin-page-treeview/lib/index.js
/***********************************************************************
 *                                                                   _
 *       _____  _                           ____  _                 |_|
 *      |  _  |/ \   ____  ____ __ ___     / ___\/ \   __   _  ____  _
 *      | |_| || |  / __ \/ __ \\ '_  \ _ / /    | |___\ \ | |/ __ \| |
 *      |  _  || |__. ___/. ___/| | | ||_|\ \___ |  _  | |_| |. ___/| |
 *      |_/ \_|\___/\____|\____||_| |_|    \____/|_| |_|_____|\____||_|
 *
 *      ================================================================
 *                 More than a coder, More than a designer
 *      ================================================================
 *
 *
 *      - Document: index.js
 *      - Author: aleen42
 *      - Description: the main entrance for page-treeview
 *      - Create Time: Apr 11st, 2016
 *      - Update Time: Apr 11st, 2016
 *
 *
 **********************************************************************/

/**
 * [replaceAll: replace all match sub-string in a string]
 * @param  {[type]} search      [description]
 * @param  {[type]} replacement [description]
 * @return {[type]}             [description]
 */
String.prototype.replaceAll = function(search, replacement) {
    var target = this;
    return target.split(search).join(replacement);
};

/**
 * [leftTrim: trim left space]
 * @return {[type]} [description]
 */
String.prototype.leftTrim = function() {
    return this.replace(/^\s+/, '');
}

/**
 * [removeMarkdown: convert markdown into plain text]
 * @type {[type]}
 */
const RemoveMarkdown = require('remove-markdown');

/**
 * [remarkable: convert markdown into a html]
 * @type {[type]}
 */
const Remarkable = require('remarkable');

/**
 * [main module]
 * @type {Object}
 */
const pageTreeview = module.exports = {
	/** Map of new style */
	book: {
		assets: './assets',
		css: [
            'style.css'
        ]
	},

    testRemove: function (str) {
        console.log(RemoveMarkdown(str));
    },

    /** Map of hooks */
    hooks: {
    	'page:before': function (page) {
            if (this.output.name != 'website') {
                return page;
            }
            
    		/**
    		 * [defaultOption: default option]
    		 * @type {Object}
    		 */
    		const defaultOption = {
                'style': 'markdown',
                'copyright': 'Copyright &#169; aleen42'
            };

    		/** if users have its option, and then combine it with default options */
			if (this.config.get('pluginsConfig')['page-treeview']) {
			// @deprecated
			// if (this.options.pluginsConfig['page-treeview']) {
				for (var item in defaultOption) {
    				/** special for copyright */
					// @deprecated
					// defaultOption[item] = this.options.pluginsConfig['page-treeview'][item] || defaultOption[item];
					defaultOption[item] = this.config.get('pluginsConfig')['page-treeview'][item] || defaultOption[item];
    			}
    		}

            var insertTreeview = '<div class="treeview__container"><div class="treeview__container-title"></div>' + new Remarkable().render(pageTreeview.initHeaders(page.content, defaultOption.style)) + '</div>';

            page.content = insertTreeview + '\n\n' + page.content;
            
    		return page;
    	}
    },

    /** Map of new blocks */
    blocks: {},

    /** Map of new filters */
    filters: {},

    /**
     * [test: tests function]
     * @param  {[type]} configs  [simulated configs]
     * @param  {[type]} contents [simulated contents]
     * @return {[type]}          [description]
     */
    test: function (configs, contents) {
        configs = configs || null;
        contents = contents || '';

    	/**
         * [defaultOption: default option]
         * @type {Object}
         */
        const defaultOption = {
            'style': 'markdown',
            'copyright': 'Copyright &#169; aleen42'
        };

        /** if users have its option, and then combine it with default options */
        if (configs) {
        // @deprecated
        // if (this.options.pluginsConfig['page-treeview']) {
            for (var item in defaultOption) {
                /** special for copyright */
                // @deprecated
                // defaultOption[item] = this.options.pluginsConfig['page-treeview'][item] || defaultOption[item];
                defaultOption[item] = configs[item] || defaultOption[item];
            }
        }

        var insertTreeview = '<div class="treeview__container"><div class="treeview__container-title"><span class="treeview__main-title">Treeview</span><span class="treeview__copyright">' + defaultOption.copyright + ' all right reserved, powered by <a href="https://github.com/aleen42" target="_blank">aleen42</a></span></div>' + new Remarkable().render(pageTreeview.initHeaders(contents, defaultOption.style)) + '</div>';

        return insertTreeview;
    },

    /**
     * [getLevelTab: generate tab for a tree with an array]
     * @param  {[type]} array [description]
     * @param  {[type]} index [description]
     * @param  {[type]} min   [description]
     * @return {[type]}       [description]
     */
    getLevelTab: function (array, index, min) {
        var ret = '';

        if (index === 0) {
            if (array[index] !== min) {
                for (var i = min; i < array[index]; i++) {
                    ret += new Array(i - min + 1).join('\t') + '- &nbsp; \n';
                }    
            }
        } else {
            if (array[index] - array[index - 1] > 1) {
                for (var i = array[index - 1] + 1; i < array[index]; i++) {
                    ret += new Array(i - min + 1).join('\t') + '- &nbsp; \n';
                }
            }
        }

        return ret + new Array(array[index] - min + 1).join('\t');
    },

    /**
     * [initHeaders: get headers of a page]
     * @param  {[type]} content [description]
     * @param  {[type]} style   [description]
     * @return {[type]}         [description]
     */
    initHeaders: function (content, style) {
        var res = [];
        var minLevel = 6;
        var levelTab = [];

        function calcLevel(str) {
            var count = 0;
            var strArr = str.split('');

            for (var i = 0; i < str.length && strArr[i] === '#'; i++) {
                count++;
            }

            return count;
        }

        var headers = [].concat(content.match(/^(#)+(\s)*(.*)\n/g)).concat(content.match(/\n(#)+(\s)*(.*)\n/g)).map(function (x) {
            if (x !== null) {
                var ret = x.replaceAll('\n', '');

                /** filter in code block and code line */
                var codeBlock = content.match(/\n```[a-zA-Z]*([\s\S]*?)```/g);
                var codeLine = content.match(/\n`[a-zA-Z]*([\s\S]*?)`/g);
                
                if (codeBlock !== null) {
                    for (var i = 0; i < codeBlock.length; i++) {
                        if (codeBlock[i].indexOf(ret) >= 0) {
                            console.log('codeBlock:' + codeBlock);
                            console.log('ret:' + ret);

                            return null;
                        }
                    }
                }

                if (codeLine !== null) {
                    for (var i = 0; i < codeLine.length; i++) {
                        if (codeLine[i].indexOf(ret) >= 0) {
                            console.log('codeLine:' + codeLine);
                            console.log('ret:' + ret);

                            return null;
                        }
                    }
                }

                ret = ret.replace(/<.*?\/>/g, '')
                    .replace(/<.*?>(.*?)<\/.*?>/g, '$1');

                var level = calcLevel(ret);

                if (level <= minLevel) {
                    minLevel = level;
                }

                levelTab.push(level);

                return ret;
            } else {
                return x;
            }
        });

        var nullCount = 0;

        for (var i = 0; i < headers.length; i++) {
            if (headers[i] === null) {
                nullCount++;
                continue;
            } else {
                var levelLength = calcLevel(headers[i]);
                var value = headers[i].replace(new Array(levelLength + 1).join('#'), '').leftTrim();
                
                var title = RemoveMarkdown(pageTreeview.filterValue(value));
                var href = pageTreeview.convertID(value);

                switch (style) {
                    case 'markdown':
                        res.push(pageTreeview.getLevelTab(levelTab, i - nullCount, minLevel) + '- [' + title + '](#' + href + ')');
                        break;
                    case 'tag':
                        res.push('<a href="#' + href + '" class="level' + levelLength + '">' + title + '</a>');
                        break;
                    default:
                        break;
                }
            }
        }

        return res.join('\n');
    },

    /**
     * [filterValue: filter value of headers]
     * @param  {[type]} str [description]
     * @return {[type]}     [description]
     */
    filterValue: function (str) {
        /** reencode of cases: [sub] title [back]() */
        var ret = str;
        var prefix = ret.match(/\[(.+)?\](?!\()+/g);

        if (prefix !== null) {
            /**
             * [A2U: ascii to unicode]
             */
            function A2U(str) {
                var reserved = '';
                var convertCh = ['[', ']'];

                for (var i = 0; i < str.length; i++) {
                    if ((i === 0 || i === str.length - 1) && convertCh.indexOf(str[i]) >= 0) {
                        reserved += '&#' + str.charCodeAt(i) + ';';
                    } else {
                        reserved += str[i];
                    }
                }

                return reserved;
            }

            ret = ret.replaceAll(prefix[0], A2U(prefix[0]));
        }

        return ret;
    },

    /**
     * [convertID: convert a markdown string into a id generated by gitbook]
     * @param  {[type]} str [description]
     * @return {[type]}     [description]
     */
    convertID: function (str) {
        var res = RemoveMarkdown(
            /** convert \t into ---- */
            str.replaceAll('\t', '----')
            /** remove “ and ” */
            .replaceAll('“', '')
            .replaceAll('”', '')
            .toLowerCase()
            .replaceAll(' ', '-')
        );

        var strArr = res.split('');

        for (var i = 0; i < strArr.length; i++) {
            /** not alpha */
            if (!(strArr[i].charCodeAt(0) >= 97 && strArr[i].charCodeAt(0) <= 122) && !(strArr[i].charCodeAt(0) === 45) && !(strArr[i].charCodeAt(0) > 126) && !(strArr[i].charCodeAt(0) >= 48 && strArr[i].charCodeAt(0) <= 57)) {
                strArr.splice(i, 1);
                i--;
            }
        }

        return strArr.join('');
    }
};
